### PR TITLE
HAPP 474 Add comments to medication records when the device has internet connection

### PR DESCRIPTION
**Pull request by rashmi-bambhania**

### CODE REVIEW

File: app/src/main/res/layout/add_comment.xml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:id="@+id/tip_comment"
+    style="@style/HealthGateway.Widget.App.TextInputLayout"
+    android:hint="@string/comment_here"
+    android:filterTouchesWhenObscured="true"
+    app:startIconDrawable="@drawable/ic_lock_solid_grey"
+    app:startIconContentDescription="@string/comment_here"
+    app:endIconContentDescription="@string/comment_here"
+    app:endIconDrawable="@drawable/ic_add_comment"
+    app:endIconMode="custom">
```
Comment (amit-metri) [inline app/src/main/res/layout/add_comment.xml:11]: little elevation is required for this view as per the screen design.

File: app/src/main/java/ca/bc/gov/bchealth/ui/comment/CommentsViewModel.kt [hunk 2]
```diff
@@ -57,6 +57,37 @@ class CommentsViewModel @Inject constructor(
         }
     }
 
+    fun addComment(parentEntryId: String, userProfileId: String?, comment: String) =
```
Comment (amit-metri) [inline app/src/main/java/ca/bc/gov/bchealth/ui/comment/CommentsViewModel.kt:60]: userProfileId is hdId, which is available at repo level. Can we remove it from here?

File: app/src/main/java/ca/bc/gov/bchealth/ui/healthrecord/medication/MedicationDetailsFragment.kt [hunk 3]
```diff
@@ -60,7 +63,10 @@ class MedicationDetailsFragment : Fragment(R.layout.fragment_medication_details)
     private fun setUpRecyclerView() {
         commentsAdapter = CommentsAdapter { parentEntryId ->
             val action = MedicationDetailsFragmentDirections
-                .actionMedicationDetailsFragmentToCommentsFragment(parentEntryId)
+                .actionMedicationDetailsFragmentToCommentsFragment(
+                    parentEntryId,
+                    userProfileId ?: ""
```
Comment (amit-metri) [inline app/src/main/java/ca/bc/gov/bchealth/ui/healthrecord/medication/MedicationDetailsFragment.kt:68]: userProfileId is hdId, which is available at repo level. Can we remove it from here?

File: app/src/main/java/ca/bc/gov/bchealth/ui/healthrecord/medication/MedicationDetailsViewModel.kt [hunk 4]
```diff
@@ -42,6 +42,7 @@ class MedicationDetailsViewModel @Inject constructor(
                     .getComments(
                         medicationWithSummaryAndPharmacyDto.medicationRecord.prescriptionIdentifier
                     )
+            var userProfileId: String? = null
```
Comment (amit-metri) [inline app/src/main/java/ca/bc/gov/bchealth/ui/healthrecord/medication/MedicationDetailsViewModel.kt:45]: userProfileId is hdId, which is available at repo level. Can we remove it from here?

File: data/src/main/java/ca/bc/gov/data/datasource/remote/model/request/CommentRequest.kt [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package ca.bc.gov.data.datasource.remote.model.request
+
+/**
+ * @author: Created by Rashmi Bambhania on 18,April,2022
+ */
+data class CommentRequest(
+    val text: String,
+    val parentEntryId: String?,
+    val userProfileId: String?,
+    val entryTypeCode: String?
```
Comment (amit-metri) [inline data/src/main/java/ca/bc/gov/data/datasource/remote/model/request/CommentRequest.kt:10]: Are we not going to use below fields as per the swagger 
"version": 0,
  "createdDateTime": "2022-04-22T08:08:29.027Z",
  "createdBy": "string",
  "updatedDateTime": "2022-04-22T08:08:29.027Z",
  "updatedBy": "string"

File: data/src/main/java/ca/bc/gov/data/datasource/remote/CommentRemoteDataSource.kt [hunk 6]
```diff
@@ -33,4 +34,30 @@ class CommentRemoteDataSource @Inject constructor(
...
+        userProfileId: String?,
+        comment: String,
+        hdid: String,
+        accessToken: String
+    ): CommentDto {
+        val commentRequest = CommentRequest(
+            text = comment,
+            parentEntryId = parentEntryId,
+            userProfileId = userProfileId,
+            entryTypeCode = "Med"
```
Comment (amit-metri) [inline data/src/main/java/ca/bc/gov/data/datasource/remote/CommentRemoteDataSource.kt:49]: We need to pass entryTypeCode from the UI layer as this code will differ for lab tests.

File: app/src/main/java/ca/bc/gov/bchealth/ui/comment/CommentsViewModel.kt [hunk 7]
```diff
@@ -57,6 +57,37 @@ class CommentsViewModel @Inject constructor(
...
+                                it.createdDateTime.toLocalDateTimeInstant()
+                            )
+                        }
+                    )
+                }
+            } catch (e: Exception) {
+                e.printStackTrace()
+                _uiState.update {
+                    it.copy(
+                        onError = true
```
Comment (amit-metri) [inline app/src/main/java/ca/bc/gov/bchealth/ui/comment/CommentsViewModel.kt:85]: onLoading to be false in case of error.

**Review by amit-metri — CHANGES_REQUESTED** [review]

File: data/src/main/java/ca/bc/gov/data/datasource/remote/model/request/CommentRequest.kt (hunk 5 above)
Comment (rashmi-bambhania) [reply data/src/main/java/ca/bc/gov/data/datasource/remote/model/request/CommentRequest.kt:10]: Not required

**Review by amit-metri — APPROVED** [review]

**Pull request merged**